tianshou/data/collector.py

```diff
@@ -242,18 +236,168 @@ def reset_buffer(self, keep_statistics: bool = False) -> None:
+        if isinstance(info_N, dict):  # type: ignore[unreachable]
             # this can happen if the env is an envpool env. Then the thing returned by reset is a dict
             # with array entries instead of an array of dicts
             # We use Batch to turn it into an array of dicts
-            self._pre_collect_info_R = _dict_of_arr_to_arr_of_dicts(self._pre_collect_info_R)  # type: ignore[unreachable]
+            info_N = _dict_of_arr_to_arr_of_dicts(info_N)  # type: ignore[unreachable]
+        return obs_NO, info_N
+
+    @abstractmethod
+    def _collect(
+        self,
+        n_step: int | None = None,
+        n_episode: int | None = None,
+        random: bool = False,
+        render: float | None = None,
+        no_grad: bool = True,
+        gym_reset_kwargs: dict[str, Any] | None = None,
+    ) -> CollectStats:
+        pass
+
+    def collect(
+        self,
+        n_step: int | None = None,
+        n_episode: int | None = None,
+        random: bool = False,
+        render: float | None = None,
+        no_grad: bool = True,
+        reset_before_collect: bool = False,
+        gym_reset_kwargs: dict[str, Any] | None = None,
+        eval_mode: bool = False,
+    ) -> CollectStats:
+        """Collect a specified number of steps or episodes.
+
+        To ensure an unbiased sampling result with the n_episode option, this function will
+        first collect ``n_episode - env_num`` episodes, then for the last ``env_num``
+        episodes, they will be collected evenly from each env.
+
+        :param n_step: how many steps you want to collect.
+        :param n_episode: how many episodes you want to collect.
+        :param random: whether to use random policy for collecting data.
+        :param render: the sleep time between rendering consecutive frames.
+        :param no_grad: whether to retain gradient in policy.forward().
+        :param reset_before_collect: whether to reset the environment before collecting data.
+            (The collector needs the initial obs and info to function properly.)
+        :param gym_reset_kwargs: extra keyword arguments to pass into the environment's
+            reset function. Only used if reset_before_collect is True.
+        :param eval_mode: whether to collect data in evaluation mode. Will
+            set the policy to training mode otherwise.
+
+        .. note::
+
+            One and only one collection number specification is permitted, either
+            ``n_step`` or ``n_episode``.
+
+        :return: The collected stats
+        """
+        # check that exactly one of n_step or n_episode is set and that the other is larger than 0
+        self._validate_n_step_n_episode(n_episode, n_step)
+
+        if reset_before_collect:
+            self.reset(reset_buffer=False, gym_reset_kwargs=gym_reset_kwargs)
+
+        policy_mode_context = in_eval_mode if eval_mode else in_train_mode
```

Comment:
Suggest to add a context manager which accepts a flag, such that this usage pattern is simplified.

Reply:
Actually, Max and I have now concluded that there is really no case where the collector should ever use the policy in train mode. The _only_ reason why it currently needs to be in train mode is that some policies have a `deterministic_eval` flag, which is True by default (e.g. SAC) and which should basically never be active when collecting data for training.

Reply:
The flag `no_grad` is also unnecessary and should be removed from `Collector.collect`.

Reply:
> Actually, Max and I have now concluded that there is really no case where the collector should ever use the policy in train mode. The only reason why it currently needs to be in train mode is that some policies have a deterministic_eval flag, which is True by default (e.g. SAC) and which should basically never be active when collecting data for training.

Max and I have resolved this. See recent commits.

Reply:
I will remove no_grad as part of this PR